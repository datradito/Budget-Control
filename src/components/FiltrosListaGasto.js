import React from 'react';
import {connect} from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByImporte, sortByFecha, setFechaFin, setFechaInicio } from '../actions/filters';

/**
 * https://github.com/airbnb/react-dates
 */

export class FiltrosListaGasto extends React.Component{
    state = {
        //Necesito seguir el estado de 'calendarFocused' como en <FormGasto/>, default es null y luego pasa a string(calendario 1 o 2)
        calendarFocused: null
    };

    //SEGUN react-dates library
    onDatesChange = ({ fechaInicio , fechaFin}) => {
        this.props.setFechaInicio(fechaInicio);
        this.props.setFechaFin(fechaFin);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ( {calendarFocused} ))
    };

    onSortChange = (e) => {
        if(e.target.value==='date'){
            this.props.sortByFecha();
        }
        else if(e.target.value==='amount'){
            this.props.sortByImporte();
        }
    };

    onTextChange =  (e) => {
        this.props.setTextFilter(e.target.value);
    };

    render() {
        return (
            <div>
            <input 
                type="text" 
                //Tomo el valor text del filters en el store
                value={this.props.filters.text} 
                //Accedo a .dispatch() desde las props porque me lo envia connect()
                onChange={this.onTextChange} />

            <select 
                value={this.props.filters.sortBy}
                onChange={ this.onSortChange }
            >
                <option value="date">Fecha</option>
                <option value="amount">Importe</option>
            </select>

            <DateRangePicker
                startDate={this.props.filters.fechaInicio} // momentPropTypes.momentObj or null,
                endDate={this.props.filters.fechaFin} // momentPropTypes.momentObj or null,
                onDatesChange = {this.onDatesChange}
                //Toma el valor del state
                focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange= {() => false }
            />
        </div>
        );
    }
};

const mapStateToProps = (state) => ({
    filters : state.filters
});

const mapDispatchToProps = () => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByFecha: () => dispatch(sortByFecha()),
    sortByImporte: () => dispatch(sortByImporte()),
    setFechaInicio: (fechaInicio) => dispatch(setFechaInicio(fechaInicio)),
    setFechaFin: (fechaFin) => dispatch(setFechaFin(fechaFin)),
});


export default connect( mapStateToProps, mapDispatchToProps)( FiltrosListaGasto ) ;