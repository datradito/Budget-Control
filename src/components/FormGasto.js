import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';


export default class FormGasto extends React.Component{
    //COMO EN OTRAS OCACIONES DEFINO SUPER PARA ACCEDER AL STATE
    constructor(props) {
        super(props);
        
        this.state = {
            //Si estoy editando, el state se llena con las props que se pasan a 
            //el componente <FormGasto/> si no, lleno el state con los valores por
            //defecto, ej: nota: props.gasto ? props.gasto.nota : '',
            descrip: props.gasto ? props.gasto.descrip : '',
            nota: props.gasto ? props.gasto.nota : '',
            importe: props.gasto ? (props.gasto.importe / 100).toString() : '',
            creadoAt: props.gasto ? moment(props.gasto.creadoAt) : moment(),
            calendarFocused: false,
            error:'',
        }
    }

    onChangeDescrip = (e) => {  
        const descrip = e.target.value;
        this.setState( () => ( { descrip } ) );
    };

    onChangeNota = (e) => {
        const nota = e.target.value;
        this.setState( () => ( { nota } ) );
    };

    onChangeImporte = (e) => {
        const importe = e.target.value;
        
        //REGULAR EXRPESSIONS, ver regex101.com
        if (!importe || importe.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState( () => ( { importe } ) );
        }
    };

    onDateChange = (creadoAt) => {
        if (creadoAt) {
            this.setState( () => ({ creadoAt }) );
    
        }
    };

    onFocusChange = ( {focused} ) => {
        this.setState(() => ({
            calendarFocused: focused
        }));
    };

    onSubmit = (e) => {
        //Prevengo refresh del explorador
        e.preventDefault();

        if(!this.state.importe || !this.state.descrip) {
            this.setState( () => (
                {error: 'Por favor ingrese descripcion e importe...'}
            ) )
        } else {
            this.setState( ()=>({ error:'' }));
            //props.onSubmit vienen desde el componente AddGastosPage pasada como props
            this.props.onSubmit({
                descrip: this.state.descrip,
                nota: this.state.nota,
                importe: parseFloat(this.state.importe, 10)*100, //CONVIERTO EL STRING importe EN NUMERO
                creadoAt: this.state.creadoAt.valueOf(), //USO FUNCION DE moment.js / timestamp
            })
        }
    };

    render () {
        return (
        <div>
            {/*Muestro error si es TRUE*/}
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit}>
                <input
                    type="text"
                    placeholder="Descripcion"
                    autoFocus
                    value={this.state.descrip}
                    onChange={this.onChangeDescrip}
                />
                <input
                    type="text"
                    placeholder="Importe"
                    value={this.state.importe}
                    onChange={this.onChangeImporte}
                />
                
                {/*https://github.com/airbnb/react-dates*/} 
                <SingleDatePicker 
                    //PROPS:
                    date= {this.state.creadoAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}

                    //SET PARA VER BOTON DE clear EN CALENDARIO, DEFAULT ES false
                    showClearDates = {() => {true}}
                    //SET PARA VER UN SOLO CALENDARIO LUEGO DE ELEGIR RANGO
                    numberOfMonths={1}
                    //SET PARA VER CALENDARIO DEL PASADO LUEGO DE ELEGIR RANGO
                    isOutsideRange={() => (false)}
                />
                
                <textarea
                    placeholder="Ingresar nota para el Gasto (opcional)"
                    value={this.state.nota}
                    onChange={this.onChangeNota}
                />
                <button>Agregar Gasto</button>
            </form>
        </div>
        )}
}