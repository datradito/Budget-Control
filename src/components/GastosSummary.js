import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectedGastos from '../selectors/gastos';
import totales_gastos from '../selectors/totales_gastos';

export const GastosSummary = ({gastosCount, gastosTotal}) => {
    const gastoWord = (gastosCount >= 2 ) ? 'gastos' : 'gasto';
    const format_gastoTotal = numeral( gastosTotal / 100 ).format('$0,0.00');
    
    return (
        <div>
            <h1>Viendo {gastosCount} {gastoWord} totalizando {format_gastoTotal}</h1>
        </div>
    );
};

const mapStateToProps = (state) => {
    const gastoFiltrado = selectedGastos(state.gastos, state.filters);
    return {
        gastosCount: gastoFiltrado.length,
        gastosTotal: totales_gastos(gastoFiltrado)
    };
};

export default connect(mapStateToProps)(GastosSummary);

/**
 * ******************************************************
 * OTRA FORMA DE CONECTAR CON EL STORE Y MANEJAR EL STATE
 * ******************************************************
 * 
 * export const GastosSummary = (props) => (
    <div>
        <h1>Total de gasto: {props.gastos.length}</h1>
        <h1>Total importe:
            {(props.gastos
                .map((gasto) => gasto.importe)
                .reduce((sum, value) => (sum + value), 0))/100
            }
        </h1>
    </div>
);

const mapStateToProps = (state) => {
    return {
        gastos: selectedGastos(state.gastos, state.filters)
    };
};
 */