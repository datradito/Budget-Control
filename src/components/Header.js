import React from 'react';
import {  NavLink } from 'react-router-dom';


const Header = () => (
    <header>
        <h1>Control de Gastos Giuliana</h1>
        <NavLink to="/" exact={true} activeClassName="is-active">Pantalla Principal</NavLink>
        <NavLink to="/crear" activeClassName="is-active">Add GastosPage Test</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help!</NavLink>
    </header>
)

export default Header;