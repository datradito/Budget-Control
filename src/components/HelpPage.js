import React from 'react';

const HelpPage = () => (
    <div>
        Ayuda
    </div>

);

export default HelpPage;