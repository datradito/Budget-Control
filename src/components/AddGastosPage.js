import React from 'react';
import { connect } from 'react-redux';
import FormGasto from './FormGasto';
import { startAddGasto } from '../actions/gastos';

export class AddGastosPage extends React.Component {
    onSubmit = (gasto) => {
        this.props.startAddGasto(gasto);
        this.props.history.push('./'); 
    };
    render() {
        return (
            <div>
            <h1>Agrega Gasto</h1>
            <FormGasto onSubmit = {this.onSubmit}/>
        </div>
        );
    }
}

//Modifico el export para poder usar spy en Jest
/**
 * https://react-redux.js.org/using-react-redux/connect-mapdispatch
 */
const mapDispatchToProps = (dispatch) => ({
    startAddGasto: (gasto) => dispatch(startAddGasto(gasto))
})
export default connect(undefined, mapDispatchToProps)(AddGastosPage);



/**
 * ********************************************************************************
 * Documento el stateless component que se utilizaba antes de configurar para tests
 * ********************************************************************************
 */

//Al utilizar connect("state")("Componente") tengo acceso a .dispatch() dsd props. Ej: 
//const Componente = (props) => (jsx)

// const AddGastosPage = (props) => (
//     <div>
//         <h1>Agrega Gasto</h1>
//        <FormGasto 
//             //La prop OnSubmit = {} recibe dsd el componente <FormGasto /> el objeto con los datos para enviar al store, 
//             onSubmit = {(gasto) => {
//                 console.log(gasto);
//                 //props.dispatch(addGasto(gasto));
//                 props.onSubmit(gasto);
//                 //El componente recibe history kmo prop a traves de react-router, ver API.
//                 //con .push('/') redirijo al dashboard page.
//                 props.history.push('./'); 
//            }}
//        />
//     </div>

// );