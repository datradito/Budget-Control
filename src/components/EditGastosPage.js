import React from 'react';
import FormGasto from './FormGasto';
import { connect } from 'react-redux';
import { editGasto, removeGasto } from '../actions/gastos';

//ACCEDO A props QUE PASA EL COMPONENTE Route 
const EditGastosPage = (props) => {
    return (
        <div>
            <FormGasto 
                gasto= {props.gasto}
                onSubmit={ (gasto) => {
                    //dispatch(action())
                    props.dispatch(editGasto(props.gasto.id, gasto));
                    //redirijo a home
                    props.history.push('/');
                    console.log('Gasto actualizado', gasto);
                }}
            />
            <button 
                    onClick = { () => {
                    props.dispatch(removeGasto({id : props.gasto.id})); 
                    //redirect:
                    props.history.push('/');}}>
                    Borrar
            </button>
        </div>
    );
};


const mapsStateToProps = (state, props) => {
    return {
        //USO props.match.params PARA ACCEDER A LOS PARAMETROS DEL COMPONENTE <Route/>
        //find() javascript al array state para obtener el obj gasto con el id buscado
        gasto: state.gastos.find((gasto) => gasto.id === props.match.params.id)
    };
};

export default connect(mapsStateToProps)(EditGastosPage);