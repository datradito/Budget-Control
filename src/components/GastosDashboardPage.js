import React from 'react';
import ListaGasto from './ListaGasto';
import FiltrosListaGasto from './FiltrosListaGasto';
import GastosSummary from './GastosSummary';

const GastosDashboardPage = () => (
    <div>
        <GastosSummary />
        <FiltrosListaGasto />
        <ListaGasto />
    </div>

);

export default GastosDashboardPage;