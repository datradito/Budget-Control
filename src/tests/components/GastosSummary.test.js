import React from 'react';
import { shallow } from 'enzyme';
import { GastosSummary } from '../../components/GastosSummary';

//beforeEach(() => wrapper = shallow(<TotalGastos />));


test('Deberia render GastosSummary', () => {
    const wrapper = shallow(<GastosSummary />);
    expect(wrapper).toMatchSnapshot();
}); 