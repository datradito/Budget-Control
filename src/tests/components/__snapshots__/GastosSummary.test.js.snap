// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deberia render GastosSummary 1`] = `
<div>
  <h1>
    Viendo 
     
    gasto
     totalizando 
    $0.00
  </h1>
</div>
`;
